File: ui-ngx/src/app/core/http/device.service.ts
///
/// Copyright © 2016-2026 The Thingsboard Authors
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///

import { Injectable } from '@angular/core';
import { createDefaultHttpOptions, defaultHttpOptionsFromConfig, RequestConfig } from './http-utils';
import { catchError, Observable, of, ReplaySubject, throwError, timeout } from 'rxjs';
import { map, switchMap } from "rxjs/operators";
import { HttpClient } from '@angular/common/http';
import { PageLink } from '@shared/models/page/page-link';
import { PageData } from '@shared/models/page/page-data';
import {
  ClaimRequest,
  ClaimResult,
  Device,
  DeviceCredentials,
  DeviceInfo,
  DeviceInfoQuery,
  DeviceSearchQuery,
  PublishTelemetryCommand,
  SaveDeviceParams
} from '@shared/models/device.models';
import { EntitySubtype } from '@shared/models/entity-type.models';
import { AuthService } from '@core/auth/auth.service';
import { BulkImportRequest, BulkImportResult } from '@shared/import-export/import-export.models';
import { PersistentRpc, RpcStatus } from '@shared/models/rpc.models';
import { ResourcesService } from '@core/services/resources.service';
import { SaveEntityParams } from '@shared/models/entity.models';

/**
 * Angular HTTP service for device CRUD, credentials, claim, bulk import, and RPC.
 *
 * <p>Wraps ThingsBoard REST endpoints under `/api/device*`, `/api/tenant/deviceInfos`, `/api/customer/{id}/device*`, and `/api/rpc/*`.
 */
@Injectable({
  providedIn: 'root'
})
export class DeviceService {

  constructor(
    private http: HttpClient,
    private resourcesService: ResourcesService
  ) { }

  
  /**
   * get device infos by query.
   *
   * @param deviceInfoQuery device info query (DeviceInfoQuery)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PageData<DeviceInfo>> observable or value
   */


  public getDeviceInfosByQuery(deviceInfoQuery: DeviceInfoQuery, config?: RequestConfig): Observable<PageData<DeviceInfo>> {
    return this.http.get<PageData<DeviceInfo>>(`/api${deviceInfoQuery.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get tenant device infos.
   *
   * @param pageLink pagination and sort parameters
   * @param type type (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PageData<DeviceInfo>> observable or value
   */


  public getTenantDeviceInfos(pageLink: PageLink, type: string = '',
                              config?: RequestConfig): Observable<PageData<DeviceInfo>> {
    return this.http.get<PageData<DeviceInfo>>(`/api/tenant/deviceInfos${pageLink.toQuery()}&type=${type}`,
      defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get tenant device infos by device profile id.
   *
   * @param pageLink pagination and sort parameters
   * @param deviceProfileId device profile id (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PageData<DeviceInfo>> observable or value
   */


  public getTenantDeviceInfosByDeviceProfileId(pageLink: PageLink, deviceProfileId: string = '',
                                               config?: RequestConfig): Observable<PageData<DeviceInfo>> {
    return this.http.get<PageData<DeviceInfo>>(`/api/tenant/deviceInfos${pageLink.toQuery()}&deviceProfileId=${deviceProfileId}`,
      defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get customer device infos.
   *
   * @param customerId customer UUID
   * @param pageLink pagination and sort parameters
   * @param type type (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PageData<DeviceInfo>> observable or value
   */


  public getCustomerDeviceInfos(customerId: string, pageLink: PageLink, type: string = '',
                                config?: RequestConfig): Observable<PageData<DeviceInfo>> {
    return this.http.get<PageData<DeviceInfo>>(`/api/customer/${customerId}/deviceInfos${pageLink.toQuery()}&type=${type}`,
      defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get customer device infos by device profile id.
   *
   * @param customerId customer UUID
   * @param pageLink pagination and sort parameters
   * @param deviceProfileId device profile id (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PageData<DeviceInfo>> observable or value
   */


  public getCustomerDeviceInfosByDeviceProfileId(customerId: string, pageLink: PageLink, deviceProfileId: string = '',
                                                 config?: RequestConfig): Observable<PageData<DeviceInfo>> {
    return this.http.get<PageData<DeviceInfo>>(`/api/customer/${customerId}/deviceInfos${pageLink.toQuery()}&deviceProfileId=${deviceProfileId}`,
      defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get device.
   *
   * @param deviceId device UUID
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Device> observable or value
   */


  public getDevice(deviceId: string, config?: RequestConfig): Observable<Device> {
    return this.http.get<Device>(`/api/device/${deviceId}`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get devices.
   *
   * @param deviceIds device ids (Array<string>)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Array<Device>> observable or value
   */


  public getDevices(deviceIds: Array<string>, config?: RequestConfig): Observable<Array<Device>> {
    return this.http.get<Array<Device>>(`/api/devices?deviceIds=${deviceIds.join(',')}`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get device info.
   *
   * @param deviceId device UUID
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<DeviceInfo> observable or value
   */


  public getDeviceInfo(deviceId: string, config?: RequestConfig): Observable<DeviceInfo> {
    return this.http.get<DeviceInfo>(`/api/device/info/${deviceId}`, defaultHttpOptionsFromConfig(config));
  }

  /** Calls ThingsBoard REST `/api/device, ...`. */

  public saveDevice(device: Device, config?: RequestConfig): Observable<Device>;
  public saveDevice(device: Device, saveParams?: SaveDeviceParams, config?: RequestConfig): Observable<Device>;
  /**
   * POST/PUT entity — save device.
   *
   * @param device device (Device)
   * @param saveParamsOrConfig save params or config (SaveDeviceParams | RequestConfig)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Device> observable or value
   */
  public saveDevice(device: Device, saveParamsOrConfig?: SaveDeviceParams | RequestConfig, config?: RequestConfig): Observable<Device> {
    return this.http.post<Device>('/api/device', device, createDefaultHttpOptions(saveParamsOrConfig, config));
  }

  /** Calls ThingsBoard REST `/api/device-with-credentials, ...`. */

  public saveDeviceWithCredentials(device: Device, credentials: DeviceCredentials, config?: RequestConfig): Observable<Device>;
  public saveDeviceWithCredentials(device: Device, credentials: DeviceCredentials, saveParams: SaveEntityParams, config?: RequestConfig): Observable<Device>;
  /**
   * POST/PUT entity — save device with credentials.
   *
   * @param device device (Device)
   * @param credentials credentials (DeviceCredentials)
   * @param saveParamsOrConfig save params or config (SaveEntityParams | RequestConfig)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Device> observable or value
   */
  public saveDeviceWithCredentials(device: Device, credentials: DeviceCredentials, saveParamsOrConfig?: SaveEntityParams | RequestConfig, config?: RequestConfig): Observable<Device> {
    return this.http.post<Device>('/api/device-with-credentials', {
      device,
      credentials
    }, createDefaultHttpOptions(saveParamsOrConfig, config));
  }

  /**
   * DELETE — delete device.
   *
   * @param deviceId device UUID
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   */

  public deleteDevice(deviceId: string, config?: RequestConfig) {
    return this.http.delete(`/api/device/${deviceId}`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get device types.
   *
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Array<EntitySubtype>> observable or value
   */


  public getDeviceTypes(config?: RequestConfig): Observable<Array<EntitySubtype>> {
    return this.http.get<Array<EntitySubtype>>('/api/device/types', defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get device credentials.
   *
   * @param deviceId device UUID
   * @param sync sync (boolean)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<DeviceCredentials> observable or value
   */


  public getDeviceCredentials(deviceId: string, sync: boolean = false, config?: RequestConfig): Observable<DeviceCredentials> {
    const url = `/api/device/${deviceId}/credentials`;
    if (sync) {
      const responseSubject = new ReplaySubject<DeviceCredentials>();
      const request = new XMLHttpRequest();
      request.open('GET', url, false);
      request.setRequestHeader('Accept', 'application/json, text/plain, */*');
      const jwtToken = AuthService.getJwtToken();
      if (jwtToken) {
        request.setRequestHeader('X-Authorization', 'Bearer ' + jwtToken);
      }
      request.send(null);
      if (request.status === 200) {
        const credentials = JSON.parse(request.responseText) as DeviceCredentials;
        responseSubject.next(credentials);
      } else {
        responseSubject.error(null);
      }
      return responseSubject.asObservable();
    } else {
      return this.http.get<DeviceCredentials>(url, defaultHttpOptionsFromConfig(config));
    }
  }

  
  /**
   * POST/PUT entity — save device credentials.
   *
   * @param deviceCredentials device credentials (DeviceCredentials)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<DeviceCredentials> observable or value
   */


  public saveDeviceCredentials(deviceCredentials: DeviceCredentials, config?: RequestConfig): Observable<DeviceCredentials> {
    return this.http.post<DeviceCredentials>('/api/device/credentials', deviceCredentials, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * make device public.
   *
   * @param deviceId device UUID
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Device> observable or value
   */


  public makeDevicePublic(deviceId: string, config?: RequestConfig): Observable<Device> {
    return this.http.post<Device>(`/api/customer/public/device/${deviceId}`, null, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * assign device to customer.
   *
   * @param customerId customer UUID
   * @param deviceId device UUID
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Device> observable or value
   */


  public assignDeviceToCustomer(customerId: string, deviceId: string,
                                config?: RequestConfig): Observable<Device> {
    return this.http.post<Device>(`/api/customer/${customerId}/device/${deviceId}`, null, defaultHttpOptionsFromConfig(config));
  }

  /**
   * unassign device from customer.
   *
   * @param deviceId device UUID
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   */

  public unassignDeviceFromCustomer(deviceId: string, config?: RequestConfig) {
    return this.http.delete(`/api/customer/device/${deviceId}`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * send one way rpc command.
   *
   * @param deviceId device UUID
   * @param requestBody request body (any)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<any> observable or value
   */


  public sendOneWayRpcCommand(deviceId: string, requestBody: any, config?: RequestConfig): Observable<any> {
    return this.http.post<any>(`/api/rpc/oneway/${deviceId}`, requestBody, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * send two way rpc command.
   *
   * @param deviceId device UUID
   * @param requestBody request body (any)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<any> observable or value
   */


  public sendTwoWayRpcCommand(deviceId: string, requestBody: any, config?: RequestConfig): Observable<any> {
    return this.http.post<any>(`/api/rpc/twoway/${deviceId}`, requestBody, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get persisted rpc.
   *
   * @param rpcId rpc id (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PersistentRpc> observable or value
   */


  public getPersistedRpc(rpcId: string, fullResponse = false, config?: RequestConfig): Observable<PersistentRpc> {
    return this.http.get<PersistentRpc>(`/api/rpc/persistent/${rpcId}`, defaultHttpOptionsFromConfig(config));
  }

  /**
   * DELETE — delete persisted rpc.
   *
   * @param rpcId rpc id (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   */

  public deletePersistedRpc(rpcId: string, config?: RequestConfig) {
    return this.http.delete<PersistentRpc>(`/api/rpc/persistent/${rpcId}`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get persisted rpc requests.
   *
   * @param deviceId device UUID
   * @param pageLink pagination and sort parameters
   * @param rpcStatus rpc status (RpcStatus)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PageData<PersistentRpc>> observable or value
   */


  public getPersistedRpcRequests(deviceId: string, pageLink: PageLink,
                                 rpcStatus?: RpcStatus, config?: RequestConfig): Observable<PageData<PersistentRpc>> {
    let url = `/api/rpc/persistent/device/${deviceId}${pageLink.toQuery()}`;
    if (rpcStatus && rpcStatus.length) {
      url += `&rpcStatus=${rpcStatus}`;
    }
    return this.http.get<PageData<PersistentRpc>>(url, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * find by query.
   *
   * @param query query (DeviceSearchQuery)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Array<Device>> observable or value
   */


  public findByQuery(query: DeviceSearchQuery,
                     config?: RequestConfig): Observable<Array<Device>> {
    return this.http.post<Array<Device>>('/api/devices', query, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * find by name.
   *
   * @param deviceName device name (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Device> observable or value
   */


  public findByName(deviceName: string, config?: RequestConfig): Observable<Device> {
    return this.http.get<Device>(`/api/tenant/devices?deviceName=${deviceName}`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * claim device.
   *
   * @param deviceName device name (string)
   * @param claimRequest claim request (ClaimRequest)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<ClaimResult> observable or value
   */


  public claimDevice(deviceName: string, claimRequest: ClaimRequest,
                     config?: RequestConfig): Observable<ClaimResult> {
    return this.http.post<ClaimResult>(`/api/customer/device/${deviceName}/claim`, claimRequest, defaultHttpOptionsFromConfig(config));
  }

  /**
   * unclaim device.
   *
   * @param deviceName device name (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   */

  public unclaimDevice(deviceName: string, config?: RequestConfig) {
    return this.http.delete(`/api/customer/device/${deviceName}/claim`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * assign device to edge.
   *
   * @param edgeId edge id (string)
   * @param deviceId device UUID
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<Device> observable or value
   */


  public assignDeviceToEdge(edgeId: string, deviceId: string,
                            config?: RequestConfig): Observable<Device> {
    return this.http.post<Device>(`/api/edge/${edgeId}/device/${deviceId}`,
      defaultHttpOptionsFromConfig(config));
  }

  /**
   * unassign device from edge.
   *
   * @param edgeId edge id (string)
   * @param deviceId device UUID
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   */

  public unassignDeviceFromEdge(edgeId: string, deviceId: string,
                                config?: RequestConfig) {
    return this.http.delete(`/api/edge/${edgeId}/device/${deviceId}`,
      defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get edge devices.
   *
   * @param edgeId edge id (string)
   * @param pageLink pagination and sort parameters
   * @param type type (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PageData<DeviceInfo>> observable or value
   */


  public getEdgeDevices(edgeId: string, pageLink: PageLink, type: string = '',
                        config?: RequestConfig): Observable<PageData<DeviceInfo>> {
    return this.http.get<PageData<DeviceInfo>>(`/api/edge/${edgeId}/devices${pageLink.toQuery()}&type=${type}`,
      defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * bulk import devices.
   *
   * @param entitiesData entities data (BulkImportRequest)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<BulkImportResult> observable or value
   */


  public bulkImportDevices(entitiesData: BulkImportRequest, config?: RequestConfig): Observable<BulkImportResult> {
    return this.http.post<BulkImportResult>('/api/device/bulk_import', entitiesData, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * get device publish telemetry commands.
   *
   * @param deviceId device UUID
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable<PublishTelemetryCommand> observable or value
   */


  public getDevicePublishTelemetryCommands(deviceId: string, config?: RequestConfig): Observable<PublishTelemetryCommand> {
    return this.http.get<PublishTelemetryCommand>(`/api/device-connectivity/${deviceId}`, defaultHttpOptionsFromConfig(config));
  }

  
  /**
   * Calls ThingsBoard REST `/api/device-connectivity/gateway-launch/*`.
   *
   * REST endpoint(s): `/api/device-connectivity/gateway-launch/*`
   *
   * @param deviceId device UUID
   * @returns Observable<any> observable or value
   */


  public downloadGatewayDockerComposeFile(deviceId: string): Observable<any> {
    return this.resourcesService.downloadResource(`/api/device-connectivity/gateway-launch/${deviceId}/docker-compose/download`);
  }

  /**
   * reboot device.
   *
   * @param deviceId device UUID
   * @param isBootstrapServer is bootstrap server (boolean)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable< observable or value
   */

  public rebootDevice(deviceId: string, isBootstrapServer: boolean, config?: RequestConfig): Observable<{
    result: string,
    msg: string
  }> {
    const rebootName = isBootstrapServer ? 'Bootstrap-Request Trigger' : 'Registration Update Trigger';
    return this.sendTwoWayRpcCommand(deviceId, {method: 'DiscoverAll'}, config).pipe(
      timeout(10000),
      switchMap((response: any) => {
        if (response.result && response.result.toUpperCase() === 'CONTENT') {
          const resourceId = isBootstrapServer ? 9 : 8;
          const resourcePath = `/1/0/${resourceId}`;
          return this.rebootTrigger(deviceId, resourcePath, config).pipe(
            map((responseReboot: any) => {
              if (responseReboot.result === 'CHANGED') {
                return {
                  result: 'SUCCESS',
                  msg: `<b>"${rebootName}"</b> - Started Successfully.`
                };
              } else {
                return {
                  result: 'ERROR',
                  msg: `<b>"${rebootName}"</b> failed:<pre>${JSON.stringify(responseReboot, null, 2)}</pre>`
                }
              }
            }),
            catchError(err =>
              of({
                result: 'ERROR',
                msg: `<b>"${rebootName}"</b> failed.<br>Error: ${err.message || err}`
              })
            )
          );
        } else {
          return of({
            result: 'ERROR',
            msg: `<b>"${rebootName}"</b> failed.<br>Bad registration device with id = ${deviceId}.<br><b>"DiscoverAll"</b> - RPC result is not "CONTENT"`
          });
        }
      }),
      catchError(err =>
        of({
          result: 'ERROR',
          msg: `<b>"${rebootName}"</b> failed.<br>Bad registration device with id = ${deviceId}.<br>Error: ${err.message || err}`
        })
      )
    );
  }

  /**
   * reboot trigger.
   *
   * @param deviceId device UUID
   * @param resourcePath resource path (string)
   * @param config optional HTTP request config (ignoreLoading, ignoreErrors, etc.)
   * @returns Observable< observable or value
   */

  private rebootTrigger(deviceId: string, resourcePath: string, config?: RequestConfig): Observable<{ result: string, msg?: string }> {
    return this.sendTwoWayRpcCommand(deviceId, {method: 'Execute', params: {id: resourcePath}}, config).pipe(
      timeout(10000),
      map(res => {
        if (res?.result?.toUpperCase() === 'CHANGED') {
          return {result: 'CHANGED'};
        } else {
          return {
            result: `${res?.result}`,
            msg: `${res?.error}`
          }
        }
      }),
      catchError(err => {
        return throwError(() => err);
      })
    );
  }
}

File: common/dao-api/src/main/java/org/thingsboard/server/dao/device/DeviceService.java
/**
 * Copyright © 2016-2026 The Thingsboard Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.thingsboard.server.dao.device;

import com.google.common.util.concurrent.ListenableFuture;
import org.thingsboard.server.common.data.Device;
import org.thingsboard.server.common.data.DeviceIdInfo;
import org.thingsboard.server.common.data.DeviceInfo;
import org.thingsboard.server.common.data.DeviceInfoFilter;
import org.thingsboard.server.common.data.DeviceProfile;
import org.thingsboard.server.common.data.DeviceTransportType;
import org.thingsboard.server.common.data.EntitySubtype;
import org.thingsboard.server.common.data.NameConflictStrategy;
import org.thingsboard.server.common.data.ProfileEntityIdInfo;
import org.thingsboard.server.common.data.device.DeviceSearchQuery;
import org.thingsboard.server.common.data.id.CustomerId;
import org.thingsboard.server.common.data.id.DeviceId;
import org.thingsboard.server.common.data.id.DeviceProfileId;
import org.thingsboard.server.common.data.id.EdgeId;
import org.thingsboard.server.common.data.id.TenantId;
import org.thingsboard.server.common.data.ota.OtaPackageType;
import org.thingsboard.server.common.data.page.PageData;
import org.thingsboard.server.common.data.page.PageLink;
import org.thingsboard.server.common.data.security.DeviceCredentials;
import org.thingsboard.server.dao.device.provision.ProvisionRequest;
import org.thingsboard.server.dao.entity.EntityDaoService;

import java.util.List;
import java.util.UUID;

/**
 * Persistence API for {@link Device} entities (credentials via {@link DeviceCredentialsService}).
 *
 * <p>Implementation: {@code dao} module. REST: {@code DeviceController} ({@code /api/device*}).
 */
public interface DeviceService extends EntityDaoService {

    /**
     * Finds device info by id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceId target device identifier
     * @return {@link DeviceInfo}
     */
    DeviceInfo findDeviceInfoById(TenantId tenantId, DeviceId deviceId);

    /**
     * Finds device by id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceId target device identifier
     * @return {@link Device}
     */
    Device findDeviceById(TenantId tenantId, DeviceId deviceId);

    /**
     * Finds device by id async.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceId target device identifier
     * @return future completing with {@link Device}
     */
    ListenableFuture<Device> findDeviceByIdAsync(TenantId tenantId, DeviceId deviceId);

    /**
     * Finds device by tenant id and name.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param name entity name (unique within tenant scope where applicable)
     * @return {@link Device}
     */
    Device findDeviceByTenantIdAndName(TenantId tenantId, String name);

    /**
     * Finds device by tenant id and name async.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param name entity name (unique within tenant scope where applicable)
     * @return future completing with {@link Device}
     */
    ListenableFuture<Device> findDeviceByTenantIdAndNameAsync(TenantId tenantId, String name);

    /**
     * Saves or persists device.
     *
     * @param device device ({@link Device})
     * @return {@link Device}
     */
    Device saveDevice(Device device);

    /**
     * Saves or persists device.
     *
     * @param device device ({@link Device})
     * @param doValidate whether to run validation before persist
     * @return {@link Device}
     */
    Device saveDevice(Device device, boolean doValidate);

    /**
     * Saves or persists device with access token.
     *
     * @param device device ({@link Device})
     * @param accessToken device access token for MQTT/HTTP connectivity
     * @return {@link Device}
     */
    Device saveDeviceWithAccessToken(Device device, String accessToken);

    /**
     * Saves or persists device with access token.
     *
     * @param device device ({@link Device})
     * @param accessToken device access token for MQTT/HTTP connectivity
     * @param nameConflictStrategy behavior when an entity with the same name already exists
     * @return {@link Device}
     */
    Device saveDeviceWithAccessToken(Device device, String accessToken, NameConflictStrategy nameConflictStrategy);

    /**
     * Saves a device with credentials the requested data.
     *
     * @param device device ({@link Device})
     * @param deviceCredentials device credentials ({@link DeviceCredentials})
     * @return {@link Device}
     */
    Device saveDeviceWithCredentials(Device device, DeviceCredentials deviceCredentials);

    /**
     * Saves a device with credentials the requested data.
     *
     * @param device device ({@link Device})
     * @param deviceCredentials device credentials ({@link DeviceCredentials})
     * @param nameConflictStrategy behavior when an entity with the same name already exists
     * @return {@link Device}
     */
    Device saveDeviceWithCredentials(Device device, DeviceCredentials deviceCredentials, NameConflictStrategy nameConflictStrategy);

    /**
     * Saves or persists device.
     *
     * @param provisionRequest provision request ({@link ProvisionRequest})
     * @param profile profile ({@link DeviceProfile})
     * @return {@link Device}
     */
    Device saveDevice(ProvisionRequest provisionRequest, DeviceProfile profile);

    /**
     * Assigns device to customer.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceId target device identifier
     * @param customerId customer to assign or filter by
     * @return {@link Device}
     */
    Device assignDeviceToCustomer(TenantId tenantId, DeviceId deviceId, CustomerId customerId);

    /**
     * Unassigns device from customer.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceId target device identifier
     * @return {@link Device}
     */
    Device unassignDeviceFromCustomer(TenantId tenantId, DeviceId deviceId);

    /**
     * Deletes device.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceId target device identifier
     */
    void deleteDevice(TenantId tenantId, DeviceId deviceId);

    /**
     * Finds devices by tenant id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<Device> findDevicesByTenantId(TenantId tenantId, PageLink pageLink);

    /**
     * Finds device infos by filter.
     *
     * @param filter filter ({@link DeviceInfoFilter})
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<DeviceInfo> findDeviceInfosByFilter(DeviceInfoFilter filter, PageLink pageLink);

    /**
     * Finds device id infos.
     *
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<DeviceIdInfo> findDeviceIdInfos(PageLink pageLink);

    /**
     * Finds profile entity id infos.
     *
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<ProfileEntityIdInfo> findProfileEntityIdInfos(PageLink pageLink);

    /**
     * Finds profile entity id infos by tenant id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<ProfileEntityIdInfo> findProfileEntityIdInfosByTenantId(TenantId tenantId, PageLink pageLink);

    /**
     * Finds devices by tenant id and type.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param type type ({@link String})
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<Device> findDevicesByTenantIdAndType(TenantId tenantId, String type, PageLink pageLink);

    /**
     * Finds device ids by tenant id and device profile id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceProfileId device profile id ({@link DeviceProfileId})
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<DeviceId> findDeviceIdsByTenantIdAndDeviceProfileId(TenantId tenantId, DeviceProfileId deviceProfileId, PageLink pageLink);

    /**
     * Finds devices by tenant id and type and empty ota package.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceProfileId device profile id ({@link DeviceProfileId})
     * @param type type ({@link OtaPackageType})
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<Device> findDevicesByTenantIdAndTypeAndEmptyOtaPackage(TenantId tenantId, DeviceProfileId deviceProfileId, OtaPackageType type, PageLink pageLink);

    /**
     * Counts devices by tenant id and device profile id and empty ota package.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceProfileId device profile id ({@link DeviceProfileId})
     * @param otaPackageType ota package type ({@link OtaPackageType})
     * @return the long result
     */
    long countDevicesByTenantIdAndDeviceProfileIdAndEmptyOtaPackage(TenantId tenantId, DeviceProfileId deviceProfileId, OtaPackageType otaPackageType);

    /**
     * Finds devices by tenant id and ids async.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceIds device ids ({@link List})
     * @return future completing with {@link List}
     */
    ListenableFuture<List<Device>> findDevicesByTenantIdAndIdsAsync(TenantId tenantId, List<DeviceId> deviceIds);

    /**
     * Finds devices by ids.
     *
     * @param deviceIds device ids ({@link List})
     * @return {@link List}
     */
    List<Device> findDevicesByIds(List<DeviceId> deviceIds);

    /**
     * Finds devices by ids async.
     *
     * @param deviceIds device ids ({@link List})
     * @return future completing with {@link List}
     */
    ListenableFuture<List<Device>> findDevicesByIdsAsync(List<DeviceId> deviceIds);

    /**
     * Deletes devices by tenant id.
     *
     * @param tenantId tenant that owns the entity or operation
     */
    void deleteDevicesByTenantId(TenantId tenantId);

    /**
     * Finds devices by tenant id and customer id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param customerId customer to assign or filter by
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<Device> findDevicesByTenantIdAndCustomerId(TenantId tenantId, CustomerId customerId, PageLink pageLink);

    /**
     * Finds devices by tenant id and customer id and type.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param customerId customer to assign or filter by
     * @param type type ({@link String})
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<Device> findDevicesByTenantIdAndCustomerIdAndType(TenantId tenantId, CustomerId customerId, String type, PageLink pageLink);

    /**
     * Finds devices by tenant id customer id and ids async.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param customerId customer to assign or filter by
     * @param deviceIds device ids ({@link List})
     * @return future completing with {@link List}
     */
    ListenableFuture<List<Device>> findDevicesByTenantIdCustomerIdAndIdsAsync(TenantId tenantId, CustomerId customerId, List<DeviceId> deviceIds);

    /**
     * Unassigns customer devices.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param customerId customer to assign or filter by
     */
    void unassignCustomerDevices(TenantId tenantId, CustomerId customerId);

    /**
     * Finds devices by query.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param query query ({@link DeviceSearchQuery})
     * @return future completing with {@link List}
     */
    ListenableFuture<List<Device>> findDevicesByQuery(TenantId tenantId, DeviceSearchQuery query);

    @Deprecated(since = "3.6.2", forRemoval = true)
    /**
     * Finds device types by tenant id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @return future completing with {@link List}
     */
    ListenableFuture<List<EntitySubtype>> findDeviceTypesByTenantId(TenantId tenantId);

    /**
     * Assigns device to tenant.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param device device ({@link Device})
     * @return {@link Device}
     */
    Device assignDeviceToTenant(TenantId tenantId, Device device);

    /**
     * Finds devices ids by device profile transport type.
     *
     * @param transportType transport type ({@link DeviceTransportType})
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<UUID> findDevicesIdsByDeviceProfileTransportType(DeviceTransportType transportType, PageLink pageLink);

    /**
     * Assigns device to edge.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceId target device identifier
     * @param edgeId edge id ({@link EdgeId})
     * @return {@link Device}
     */
    Device assignDeviceToEdge(TenantId tenantId, DeviceId deviceId, EdgeId edgeId);

    /**
     * Unassigns device from edge.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param deviceId target device identifier
     * @param edgeId edge id ({@link EdgeId})
     * @return {@link Device}
     */
    Device unassignDeviceFromEdge(TenantId tenantId, DeviceId deviceId, EdgeId edgeId);

    /**
     * Finds devices by tenant id and edge id.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param edgeId edge id ({@link EdgeId})
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<Device> findDevicesByTenantIdAndEdgeId(TenantId tenantId, EdgeId edgeId, PageLink pageLink);

    /**
     * Finds devices by tenant id and edge id and type.
     *
     * @param tenantId tenant that owns the entity or operation
     * @param edgeId edge id ({@link EdgeId})
     * @param type type ({@link String})
     * @param pageLink pagination and sort parameters
     * @return {@link PageData}
     */
    PageData<Device> findDevicesByTenantIdAndEdgeIdAndType(TenantId tenantId, EdgeId edgeId, String type, PageLink pageLink);

}
